feat: adds a ramps banner to home pages when users have zero balance

georgeweiler opened the pull request:
## **Description**

This PR adds a new component: `<RampsCard />` which is meant to replace the `<AssetListConversionButton/>` component.

The `AssetListConversionButton` has been removed. Insteard, a "receive tokens" link and the RampsCard have been added.

The RampsCard component will appear when the user has a zero native balance on the connected network and is meant to drive users toward the pdapp for onramping.

There are 3 variants of the `<RampsCard>` for each of the 3 home page tabs: `Tokens`, `NFTs` and `Activity`. These variants have different designs but equal functionality.


[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23352?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1.  Connect to an account with `0` balance and verify that the ramps card is appearing
2. Check each tab to verify the variants change the text, color and imagery of the ramps card.
3. Change network and verify that the copy changes to the native token of that network
4. Verify that each ramps card directs to the pdapp onramping feature

## **Screenshots/Recordings**

Designs:
<img width="357" alt="Screenshot 2024-03-06 at 9 46 55 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/72e5da0e-429f-4886-9883-be6e2faaec1d">
<img width="354" alt="Screenshot 2024-03-06 at 9 46 51 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/88405f28-22ae-439d-a3cb-ba17a453958b">
<img width="362" alt="Screenshot 2024-03-06 at 9 46 42 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/0ae7bbb0-414a-460e-a1fd-c9d83be96900">



Storybook:

<img width="1243" alt="Screenshot 2024-03-06 at 9 51 02 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/7b3a0ffb-2b4f-4888-86e4-4cbd4bc1156b">

### **Before**
<img width="372" alt="Screenshot 2024-03-06 at 9 53 58 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/6f2b8771-45a3-43ad-86c7-56d72a9a188e">

### **After**
<img width="363" alt="Screenshot 2024-03-06 at 9 58 13 AM" src="https://github.com/MetaMask/metamask-extension/assets/15269424/7388777a-8dc9-4848-9c7b-8d0b78d5b837">



![Screenshot 2024-03-19 at 12 42 06 PM](https://github.com/MetaMask/metamask-extension/assets/15269424/3549c187-4ff1-48fe-854a-4b01f9326b32)



## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

georgeweiler committed:
feat(ramps): adds a new ramps card component with gradient background images

georgeweiler committed:
feat(ramps): adds illustrations to ramps card, installs them on home pages

georgeweiler committed:
feat(ramps): adds RampsCard to the home page tabs, fixes styling, rename images

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
chore(ramp): removes AssetListConversionButton

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
refactor(ramps): use css to apply gradients instead of a png image

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
feat(ramps): adds analytics tracking to the ramps card

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
chore(ramps): fixes linting issue

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
chore(ramps): removes unused locales

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
feat: updates unit test to handle new ramps card

georgeweiler committed:
chore(ramps): removes .only

georgeweiler committed:
chore(ramps): updates test copy

georgeweiler committed:
chore(ramps): fixes nft-tab test and include case for the ramps card

georgeweiler committed:
chore: fixes nft tab test

georgeweiler committed:
chore: fix linting error

georgeweiler committed:
chore: fixes mock store for nft-tab test

georgeweiler committed:
test: removes the nft conversion banner test

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
feat: updates dependencies for EmptyBuyBannerDisplayed event effect

georgeweiler committed:
feat: update display logic for the ramps card on the NFT tab

legobeat reviewed (changes requested):
The Receive action seems to be missing? It doesn't seem appropriate to remove direct access to it from the dashboard for the sake of prominence of the Buy action?

If it's actually decided to remove the current Receive button, at minimum would seem appropriate to have a corresponding button in the above section of blue circles, next to Send?

georgeweiler committed:
feat: adds receive button to asset list when user balance is zero

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
feat(ramps): adds chainId param to the portfolio buy url (#23572)

## **Description**

Adds `chainId` param to the portfolio buy URL for deeplinking. Now the
portfolio dapp will be able to auto-select the native token for the
connected extension network when the user clicks a "buy" button.


[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23572?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Click a buy button and confirm that `chainId` is present in the URL
2. Switch networks and verify that the `chainId` param has been updated 

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it
is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
- [ ] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler commented:
Here is a video to show how the `chainId` param is used when clicking the new buy button

https://www.loom.com/share/4e34b58e511b40d687c9689ed82b5ec7

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
chore(ramps): adjust nft tab test to check for eth balance

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
chore(ramps): add token list to nft test store

kumavis commented:
I don't understand the context of how we got to the current state but I would expect a basic feature like receive to be a blue chip up next to send. link below the fold is a significant ux downgrade

georgeweiler committed:
Merge branch 'develop' into ramps-card

lorenzosantos commented:
Hi @kumavis, I appreciate your suggestion to add more "blue chip" buttons to the primary nav. However, that's outside the scope of this PR, but it's a good point for @hesterbruikman and the Wallet UX team to consider for future updates.

In this PR, we're focusing on enhancing the experience for users with zero balances by introducing a contextual 'Buy' CTA that smartly disappears once the account is funded.

Nevertheless, our PR does consider the receive functionality by adding a _persistent_ button above import tokens. Furthermore, users can conveniently copy their account address to the clipboard via the _sticky_ header nav. I suspect most users copy their account address on their desktops when seeking tokens.

**Please take a moment to review the contextual Buy CTA changes and sidebar the 'blue chip' buttons design ideas (a follow-up with Hester/Wallet UX would be great).**

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
feat(ramps): adds additional metamask entry param options when opening buy url in pdapp

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler commented on ui/components/app/nfts-tab/nfts-tab.js:
```diff
@@ -88,137 +104,112 @@ export default function NftsTab() {
-            ///: END:ONLY_INCLUDE_IF
-          }
-          <Box
-            padding={12}
```
FYI, git is confused at these changes because of the indent and the diff is incorrect. The JSX in this file changed from:

```
 <Box className="nfts-tab">
   <All other content />
 </Box>
```
to this:

```
<>
  <RampsCard/>
  <All other content />
</>
```

The only exception is the removal of the <AssetListConversionButton/>

wachunei replied:
Hiding whitespace changes helps a lot here (or adding this query param to the url `w=1`)

georgeweiler committed:
Merge branch 'develop' into ramps-card

wachunei commented on app/_locales/en/messages.json:
```diff
@@ -1837,6 +1838,12 @@
+    "message": "Fund your wallet"
+  },
+  "fundYourWalletDescription": {
+    "message": "Get started by adding some $1 to your wallet."
```
let's add the same description of `"buyToken"` here (I'm assuming `$1` is the symbol)

wachunei commented on app/_locales/en/messages.json:
```diff
@@ -1920,6 +1927,12 @@
   "genericExplorerView": {
     "message": "View account on $1"
   },
+  "getStartedWithNFTs": {
+    "message": "Get $1 to buy NFTs"
+  },
+  "getStartedWithNFTsDescription": {
+    "message": "Get started with NFTs by adding some $1 to your wallet."
+  },
```
Same as https://github.com/MetaMask/metamask-extension/pull/23352/files#r1559927987 (let's add a description)

georgeweiler replied:
[6656548](https://github.com/MetaMask/metamask-extension/pull/23352/commits/6656548615587d0876303b38f85d3630367963aa)

wachunei commented on app/_locales/en/messages.json:
```diff
@@ -4787,6 +4800,12 @@
   "stake": {
     "message": "Stake"
   },
+  "startYourJourney": {
+    "message": "Start your journey with $1"
+  },
+  "startYourJourneyDescription": {
+    "message": "Get started with web3 by adding some $1 to your wallet."
+  },
```
Same as https://github.com/MetaMask/metamask-extension/pull/23352/files#r1559927987 (let's add a description)

georgeweiler replied:
[6656548](https://github.com/MetaMask/metamask-extension/pull/23352/commits/6656548615587d0876303b38f85d3630367963aa)

wachunei commented on ui/components/app/asset-list/asset-list.js:
```diff
@@ -175,53 +119,7 @@ const AssetList = ({ onClickAsset }) => {
-          ) : null}
-        </Box>
-      ) : null}
+      {shouldShowBuy && <RampsCard variant={RAMPS_CARD_VARIANT_TYPES.TOKEN} />}
```
Two comments here:
1. Prefer [ternary over conditional rendering](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx), as the deleted code was 
2. Previous code was fencing the ramp code to be included only for `build-main,build-beta,build-flask`, we should keep the behavior

```suggestion
      {
        ///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
        shouldShowBuy ? (
          <RampsCard variant={RAMPS_CARD_VARIANT_TYPES.TOKEN} />
        ) : null
        ///: END:ONLY_INCLUDE_IF
      }
```

This would need to be paired with the import as well:
```ts
///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
import {
  RAMPS_CARD_VARIANT_TYPES,
  RampsCard,
} from '../../multichain/ramps-card/ramps-card';
///: END:ONLY_INCLUDE_IF
```

georgeweiler replied:
[40eb114](https://github.com/MetaMask/metamask-extension/pull/23352/commits/40eb114da411c8685f3f6e32479c388723bc032e)

wachunei commented on ui/components/app/nfts-tab/nfts-tab.js:
```diff
@@ -88,137 +104,112 @@ export default function NftsTab() {
-                  }}
-                />
+    <>
+      {showRampsCard && <RampsCard variant={RAMPS_CARD_VARIANT_TYPES.NFT} />}
```
Same comment as before. 
1. Prefer ternary and null as previous code was
2. the `showNftBanner` ternary was fenced by `///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)`, we should keep the behavior here

```suggestion
      {
        ///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)
        showRampsCard ? (
          <RampsCard variant={RAMPS_CARD_VARIANT_TYPES.NFT} />
        ): null
        ///: END:ONLY_INCLUDE_IF
      }
```

This should be applied to the imports as well.

georgeweiler replied:
[40eb114](https://github.com/MetaMask/metamask-extension/pull/23352/commits/40eb114da411c8685f3f6e32479c388723bc032e)

wachunei commented on ui/components/app/transaction-list/transaction-list.component.js:
```diff
@@ -205,95 +225,105 @@ export default function TransactionList({
-              .slice(0, limit)
-              .map((dateGroup) => {
+    <>
+      {showRampsCard && (
+        <RampsCard variant={RAMPS_CARD_VARIANT_TYPES.ACTIVITY} />
+      )}
```
Same as previous comments: https://github.com/MetaMask/metamask-extension/pull/23352/files?diff=split&w=1#r1559956765

georgeweiler replied:
[40eb114](https://github.com/MetaMask/metamask-extension/pull/23352/commits/40eb114da411c8685f3f6e32479c388723bc032e)

wachunei commented on ui/hooks/experiences/useRamps.test.js:
```diff
@@ -1,46 +1,108 @@
     });
   });
+
+  ['0x1', '0x38', '0xa'].forEach((mockChainId) => {
+    it(`should open the buy crypto URL with the currently connected chain ID of ${mockChainId}`, () => {
```
nit: use [`it.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) instead

georgeweiler replied:
[1b42efd](https://github.com/MetaMask/metamask-extension/pull/23352/commits/1b42efdfda01485ca81a5c121f16c4b887198e92)

wachunei reviewed (changes requested):
Left a couple of comments regarding conditional rendering and code fencing

georgeweiler committed:
Merge branch 'develop' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
feat(ramps): adds descriptions to messages.json changes that use a $1 variable

georgeweiler committed:
feat(ramps): update ramps card rendering to consider conditionals and feature flags

georgeweiler committed:
chore(ramps): refactor test to use it.each instead of looping through the array of chain ids

bkirb commented:
@georgeweiler LGTM for Ramps QA ✅ I verified the banners and the functionality

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler committed:
Merge branch 'develop' into ramps-card

danjm commented on ui/components/app/asset-list/asset-list.js:
```diff
@@ -120,48 +106,8 @@ const AssetList = ({ onClickAsset }) => {
   const balanceIsZero = Number(totalFiatBalance) === 0;
   const isBuyableChain = useSelector(getIsBuyableChain);
   const shouldShowBuy = isBuyableChain && balanceIsZero;
```
@georgeweiler This `shouldShowBuy` variable is causing linting to fail on MMI builds https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/76477/workflows/4f9e8d7b-790e-47a4-9a7b-57cc1013528e/jobs/2639253

> scripts:core:dist:standardEntryPoints: Error: MetaMask build: Lint errors encountered for transformed file "/home/circleci/project/ui/components/app/asset-list/asset-list.js":

>    no-unused-vars
 >   'shouldShowBuy' is assigned a value but never used
 
 The declaration needs to be inside a `///: BEGIN:ONLY_INCLUDE_IF(build-main,build-beta,build-flask)` code fence

klejeune committed:
feat(ramps): fix lint

danjm committed:
Fix MMI lint issues

klejeune commented:
I have read the CLA Document and I hereby sign the CLA

georgeweiler committed:
chore: expand code fences for code added in this pr

georgeweiler committed:
Merge branch 'ramps-card' of https://github.com/MetaMask/metamask-extension into ramps-card

georgeweiler committed:
fix: fix fencing typo

georgeweiler committed:
Merge branch 'develop' into ramps-card

georgeweiler commented:
The remaining two jobs that are failing are `Sonar / SonorCloud` and `ci/circleci: test-e2e-chrome-mmi` but I don't believe these changes are at fault.
